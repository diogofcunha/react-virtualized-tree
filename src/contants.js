export const UPDATE_TYPE = {
  ADD: 0,
  DELETE: 1,
  UPDATE: 2,
  MOVE_AS_CHILDREN: 3,
  MOVE_AS_SIBLING: 4
}