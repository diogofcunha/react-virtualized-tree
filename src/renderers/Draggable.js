import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import { Renderer } from '../shapes/rendererShapes';
import { UPDATE_TYPE } from '../contants';

class Draggable extends React.PureComponent {
  static propTypes = {
    ...Renderer
  }

  state = {
    draggingOver: false,
    draggingOverElement: false,
  }

  static contextTypes = {
    handleDrag: PropTypes.func.isRequired,
    handleDrop: PropTypes.func.isRequired,
    fromNode: PropTypes.object
  };

  constructor(props, context) {
    super(props, context);

    if (!context.handleDrag || !context.handleDrop) {
      throw new Error('You are trying to use Draggable renderer without the container');
    }
  }

  handleContainerDragOver = (draggingOver) => {
    this.setState({ draggingOver });
  }

  handleDragOverElement = (draggingOverElement) => {
    this.setState({ draggingOverElement });
  }

  handleDrop = dropType => e => {
    const { node } = this.props;
    
    this.context.handleDrop(node, dropType);
    e.stopPropagation();
  }

  render() {
    const containerClassName = classNames('dropable', { droping: this.state.draggingOver });
    const draggableElementClassName = classNames('draggable', { dragOverElement: this.state.draggingOverElement });
  
    return (
      <div style={{ height: 21 }}
        className={containerClassName}
        onDragOver={e => {
          e.preventDefault();
          this.handleContainerDragOver(true)
        }}
        onDragLeave={e => this.handleContainerDragOver(false)}
        onDrop={this.handleDrop(UPDATE_TYPE.MOVE_AS_SIBLING)}
      >
        <span
          style={{ height: 19 }}
          className={draggableElementClassName}
          draggable
          onDragOver={e => {
            this.handleDragOverElement(true);
            e.preventDefault();
            e.stopPropagation();
          }}
          onDragLeave={e => this.handleDragOverElement(false)}
          onDrag={e => {
            e.dataTransfer.effectAllowed = 'move';
            this.context.handleDrag(this.props.node); 
          }}
          onDrop={this.handleDrop(UPDATE_TYPE.MOVE_AS_CHILDREN)}>
            { this.props.children }
        </span>
        <div>
        </div>
      </div>);
  }
}

export default Draggable;