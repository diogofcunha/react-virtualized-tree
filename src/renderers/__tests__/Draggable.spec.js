import React from 'react';
import { shallow } from 'enzyme';

import Draggable from '../Draggable';
import { UPDATE_TYPE } from '../../contants';

describe('renderers Draggable', () => {
  const getFakeEvent = () => ({ preventDefault: jest.fn(), stopPropagation: jest.fn() });

  const setup = (extraProps = {}, extraContext = {}, state) => {
    const baseProps = {
      onChange: jest.fn(),
      node: {
        id: 1,
        name: 'Node 1',
        state: {},
        deepness: 0,
        children: [{}],
        parents: [ 2, 3 ] 
      }
    }

    const baseContext = {
      handleDrag: jest.fn(),
      handleDrop: jest.fn()
    };

    const context = { ...baseContext, ...extraContext };

    const props = { ...baseProps, ...extraProps };
    const wrapper = shallow(<Draggable {...props} />, { context })

    if (state) {
      wrapper.setState(state)
    }

    return {
      props,
      wrapper,
      context,
      containerWrapper: wrapper.first(),
      draggableElementWrapper: wrapper.findWhere(p => p.props().draggable)
    };
  };

  it('should throw an error when context is not supplied', () => {
    const componentRender = () => setup({}, { handleDrag: undefined, handleDrop: undefined });
    
    expect(componentRender).toThrowErrorMatchingSnapshot();
  });

  describe('container', () => {
    describe('styling', () => {
      it('when not dragging over, should not have droping class', () => {
        const { containerWrapper } = setup();

        expect(containerWrapper.hasClass('dropable')).toBe(true);
        expect(containerWrapper.hasClass('droping')).toBe(false);
      });

      it('when dragging over, should have droping class', () => {
        const { containerWrapper, wrapper } = setup({}, {}, { draggingOver: true });

        expect(containerWrapper.hasClass('dropable')).toBe(true);
        expect(containerWrapper.hasClass('droping')).toBe(true);
      });
    });

    it('drag over should set draggingOver state to true and prevent default', () => {
      const { wrapper, containerWrapper } = setup();
      const event = getFakeEvent();

      containerWrapper.simulate('dragOver', event);

      expect(wrapper.state().draggingOver).toBe(true);
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('drag leave should set draggingOver state to false', () => {
      const { wrapper, containerWrapper } = setup();

      containerWrapper.simulate('dragLeave');

      expect(wrapper.state().draggingOver).toBe(false);
    });

    it('drop should call injected handleDrop with the expected params and stop propagation', () => {
      const { containerWrapper, context, props } = setup();
      const event = getFakeEvent();

      containerWrapper.simulate('drop', event);

      expect(context.handleDrop).toHaveBeenCalledWith(props.node, UPDATE_TYPE.MOVE_AS_SIBLING);
      expect(event.stopPropagation).toHaveBeenCalled();
    });
  });

  describe('draggable element', () => {
    it('dragOver should prevent default and stop propagation', () => {
      const { draggableElementWrapper } = setup();
      const event = getFakeEvent();

      draggableElementWrapper.simulate('dragOver', event);

      expect(event.stopPropagation).toHaveBeenCalled();
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('drag should call injected handleDrag with the expected params', () => {
      const { draggableElementWrapper, context, props } = setup();

      draggableElementWrapper.simulate('drag');

      expect(context.handleDrag).toHaveBeenCalledWith(props.node);
    });

    it('drop should call injected handleDrop with the expected params and stop propagation', () => {
      const { draggableElementWrapper, context, props } = setup();
      const event = getFakeEvent();

      draggableElementWrapper.simulate('drop', event);

      expect(context.handleDrop).toHaveBeenCalledWith(props.node, UPDATE_TYPE.MOVE_AS_CHILDREN);
      expect(event.stopPropagation).toHaveBeenCalled();
    });
  });
});