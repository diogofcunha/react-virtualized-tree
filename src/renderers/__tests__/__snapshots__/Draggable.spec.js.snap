// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderers Draggable should throw an error when context is not supplied 1`] = `"You are trying to use Draggable renderer without the container"`;
